### PR TITLE
Remove usage of connectPreserveOnBackOption.

### PR DESCRIPTION
**Pull request by jainkuniya**
Removed from `MainNavBar`, `ModalSearchNavBar`, where it will always be true.
Currently WIP, need to handle cases where it will be false in `ModalNavBar`. (Also a rebase)

Edit: Completed.

### CODE REVIEW

File: src/nav/ModalNavBar.js [hunk 1]
```diff
@@ -76,11 +78,4 @@ class ModalNavBar extends PureComponent<Props> {
...
 
-export default connect(
-  (state: GlobalState) => ({
-    canGoBack: getCanGoBack(state),
-  }),
-  null,
-  null,
-  connectPreserveOnBackOption,
-)(ModalNavBar);
+export default connect((state: GlobalState) => ({}), null, null)(ModalNavBar);
```
Comment (borisyankov) [inline src/nav/ModalNavBar.js]: `(state: GlobalState) => ({}), null, null)` is that part even needed?
If there is no `dispatch` even the `connect` is not needed.

**Comment by gnprice** [comment]
Thanks @jainkuniya !

> `<ModalNavBar dispatch={dispatch} title={title} />`

I'm not a fan of passing `dispatch` as an explicit prop like this. In general I'd rather either:
* get `dispatch` from a `connect` call -- even if that's the only thing `connect` does; or
* pass specific callbacks to do the specific things the child component needs to do.

In this case, I think a callback prop like `onBack` would be a good choice. Then the parent can have something like `onBack = () => { this.props.dispatch(navigateBack()); }`.

---

Zooming out: how are we going to have this code decide whether the back/Up button should be shown?

Maybe the first question: are there any screens at all where we use `Screen` and *don't* want to show it, i.e. where `canGoBack` should be false?

I am not sure there are. I think the only screen where that should be false may be `MainScreenWithTabs`... which doesn't use `Screen`, so this code doesn't get involved.

If that's right, we can effectively just hard-code `canGoBack = true` throughout these components. (Which means we remove `canGoBack` completely, replacing it with `true` and then simplifying the code accordingly.)

**Comment by gnprice** [comment]
Later, if we need to revive some logic resembling `getCanGoBack`: starting with react-navigation v2 there is a direct way to do that [from the component's props](https://reactnavigation.org/docs/en/2.x/navigation-prop.html#dangerouslygetparent-get-parent-navigator), specifically `props.navigation`. Which would be fine; the important thing is it needs to come from the props.

**Comment by gnprice** [comment]
Thanks @jainkuniya !

I agree about those two screens.

I like the first commit; I plan to merge it in a minute.

Starting at the second commit:
* "It is hard to image a case [where it's false]" is not very reassuring. :smile: Can you give an explanation why it definitely will always be true?
* Why keep the `connectPreserveOnBackOption`, and why have a `canGoBack` defaultProp?

Third commit:
* It's not totally clear to me we weren't using this; certainly we seem to look at `state` in `render`, and use it to decide what to show. Also you're refactoring the API to change the type of a prop. Maybe this would be easier to explain as a couple of separate commits?

Fourth commit (ModalNavBar):
* What @borisyankov said above. :slightly_smiling_face: 

Fifth commit ("flow, modalNavBar"):
* Same; why the `connect` at all?
* Also, it sounds like Flow was broken at the previous commit? Remember, every commit should leave things as non-broken.

Last few commits:
* It looks like these are adding back functionality that was broken in previous commits. Every commit should leave things as non-broken. Maybe add this optional prop earlier on.

**Comment by gnprice** [comment]
And merged that first commit, as c31a1c50a.

**Comment by jainkuniya** [comment]
> Can you give an explanation why it definitely will always be true?

It is hard to imagine a case when `canGoBack` will be false for this
component, so the selector is almost redundant for this. This modal is
used in search screens, where we need to show search input in the
navBar. As per our current app flow, there is always a possibility of
going back from a search screen. So we always need to show back icon
in this modal.

>It's not totally clear to me we weren't using this;

Before we were using different type of navBar, there was a search icon
in the right of navBar and on clicking this icon we were showing
search input in the navBar else screen name was shown. But now we are
not using this style. Now in all screens, by default searchInput is
shown. So remove this code of toggling search state.

>it looks like these are adding back functionality...

Now on every commit is is working, i.e not broken :)

File: src/nav/ModalNavBar.js (hunk 1 above)
Comment (borisyankov) [reply src/nav/ModalNavBar.js]: What I meant was that you don't need to pass the three empty parameters.
```js
export default connect((state: GlobalState) => ({}), null, null)(ModalNavBar);
```
can become
```js
export default connect()(ModalNavBar);
```

Comment (jainkuniya) [reply src/nav/ModalNavBar.js]: Cool! got it 👍

File: src/nav/ModalNavBar.js [hunk 2]
```diff
@@ -4,17 +4,15 @@ import { connect } from 'react-redux';
...
 import Label from '../common/Label';
-import { getCanGoBack } from '../selectors';
 import NavButton from './NavButton';
 import { navigateBack } from '../actions';
-import { connectPreserveOnBackOption } from '../utils/redux';
 
 type Props = {|
   dispatch: Dispatch,
-  canGoBack: boolean,
+  canGoBack?: boolean,
```
Comment (borisyankov) [inline src/nav/ModalNavBar.js]: This should not be optional because you always pass a value thanks to `defaultProps`

**Comment by borisyankov** [comment]
@gnprice I did a review. Besides the single comment I made, I think the PR is good to merge.

File: src/nav/ModalNavBar.js (hunk 2 above)
Comment (jainkuniya) [reply src/nav/ModalNavBar.js]: Done 👍

**Comment by gnprice** [comment]
Thanks @jainkuniya !

> refactor: Remove unused code from `ModalSearchNavbar`.

The new explanation is helpful, thanks.

This commit is still doing several different things, and some of them don't match the commit message. In particular, it adds an extra condition `&& searchBarOnChange` in `Screen.js`. If that prop is missing, do you really want to make a `ModalNavBar` instead? Why?

The explanation also doesn't quite say why this code really is dead. "By default" is not very reassuring. :-) I think the real point is that at the only call site for this component, we don't pass these three props -- so in fact they are always `undefined`, and that leads to all these simplifications.

> modalSeachNavbar: Remove dependency of `getCanGoBack`.

Again "hard to imagine" and "almost redundant" are not very reassuring. :wink: The thing I really want to know is that the value is *always* true, because we only ever use this on screens where it's always true.

> modalNavBar: Remove the dependency of `getCanGoBack`.

Again, why does `ModalNavBar` have a default value for `canGoBack`? Or for `title`?

This commit is also doing two very different things -- these would be better as separate commits:
* It moves the determination of `canGoBack` from `ModalNavBar` up to its caller, `Screen`. This is a straightforward plumbing change -- it's a refactor that doesn't require any careful thinking about the app's UI to understand and verify.
* It moves the determination of `canGoBack` from `Screen` up to its callers. This requires thinking globally about the entire app's interface, and trying to be sure not to miss some other screen that could possibly need to not have a back icon.

The first of these two could just move the `getCanGoBack` call to `Screen`. Then the second one would remove that call, in favor of having the default and passing the exceptional values down from those two screens.

--

Looking back at these, I think two of my comments can be addressed together:
* For both kinds of `*NavBar` component, make `canGoBack` a prop passed by the caller; make `Screen` use `getCanGoBack`.
* Then, replace that last `getCanGoBack` with passing `canGoBack` from the callers of `Screen` (with a default). I think it'll happen to be the case that the only places we pass `false` there will be places we don't pass `search` as true... but that's kind of by accident.

I also don't feel 100% confident yet that these two are the only two screens that should be passing `canGoBack={false}`. Can you explain why it's really only those two screens?

**Comment by borisyankov** [comment]
> I also don't feel 100% confident yet that these two are the only two screens that should be passing canGoBack={false}. Can you explain why it's really only those two screens?

There are two things here.

First, optimally we don't need to reason about what Screen can go back, as these components are in the root of the Stack Navigator and will have a `navigation` passed to them. We can use similar logic to what we have in Redux, or something ready-made and will look something like `this.props.navigation.canGoBack()` (in this case, `canGoBack` is pseudo code, needs some digging in the documentation).

But second, it isn't too hard to determine what screens can be navigated back from.
We have two main states:
* logged in to a realm - the main screen is `MainScreenWithTabs` and it does not have a back button
* login flow - this can reliably be analyzed as it includes only a few screens. I think @jainkuniya analysis here is conclusive and only these screens really can not have a 'back' button

**Comment by gnprice** [comment]
> First, optimally we don't need to reason about what Screen can go back, as these components are in the root of the Stack Navigator and will have a `navigation` passed to them. We can use similar logic to what we have in Redux, or something ready-made and will look something like `this.props.navigation.canGoBack()`

Hmm, I like this idea! Very much agree that it'd be preferable not to have to reason about which screens can go back.


> * logged in to a realm - the main screen is `MainScreenWithTabs` and it does not have a back button
> * login flow - this can reliably be analyzed as it includes only a few screens. I think @jainkuniya analysis here is conclusive and only these screens really can not have a 'back' button

Well, I guess I don't yet see an analysis of the login flow, only a conclusion :slightly_smiling_face: . What are all the different screens in the login flow, and why is that a complete list?

Or, what are all the different ways we can land on a screen without navigating in a way we should be able to go back from, and why is that a complete list?

**Comment by jainkuniya** [comment]
> What are all the different screens in the login flow, and why is that a complete list?

Welcome Screen, RealmScreen, AuthScreen, PasswordAuthAcreen, DevAuthScreen.

There are 3 possibility:
* User freshly opens app, after install. Then welcome screen will be at the root of the navigation till user is logged in. 
* After logging/user opens app (already login case): `MainScreenWithTabs` will be at root (Main). 
* User logs out/opens app which logged out from all out: Then `AccountPickScreen` will be at the root of the navigation.

So we only need to handle 1 & 3 case, where user can not go back and don't want to show back icon in nav bar.

cc @gnprice

**Comment by gnprice** [comment]
And merged! :confetti_ball: 

I made some edits to the branch -- primarily
* splitting up some commits
* adding more reasoning on a couple of commit messages
* adding a couple of extra commits to clarify some aspects.

Do take a look at the merged version; it's 6c7fffc76..cd7b25757.

**Pull request merged**